share/vss/vss.go

```diff
@@ -0,0 +1,579 @@
+// distributed to the complainer.
+type Justification struct {
+	Complaint *Complaint
+	Deal      *Deal
+}
+
+// Approval is a message that is sent if a verifier approves the Deal he
+// received from the Dealer.
+type Approval struct {
+	SessionID []byte
+	Index     uint32
+	Signature []byte
+}
+
+// NewDealer returns a Dealer capable of leading the secret sharing scheme. It
+// does not have to be trusted by other Verifiers. The security parameter t is
+// the number of shares required to reconstruct the secret. It must be between
+// (len(verifiers) + 1)/2 <= t <= len(verifiers).
+func NewDealer(suite abstract.Suite, longterm, secret abstract.Scalar, verifiers []abstract.Point, r cipher.Stream, t int) (*Dealer, error) {
+	d := &Dealer{
+		suite:     suite,
+		long:      longterm,
+		secret:    secret,
+		verifiers: verifiers,
+	}
+	if !validT(t, verifiers) {
+		return nil, fmt.Errorf("dealer: t %d invalid.", t)
+	}
+
+	d.t = t
+
+	H := deriveH(d.suite, d.verifiers)
+	f := share.NewPriPoly(d.suite, d.t, d.secret, r)
+	g := share.NewPriPoly(d.suite, d.t, nil, r)
+	pub := d.suite.Point().Mul(nil, d.long)
+
+	// F = coeff * B
+	F := f.Commit(d.suite.Point().Base())
+	// G = coeff * H
+	G := g.Commit(H)
+
+	C, err := F.Add(G)
+	if err != nil {
+		return nil, err
+	}
+	_, commitments := C.Info()
+
+	d.sessionID, err = sessionID(d.suite, pub, d.verifiers, commitments, d.t)
+	if err != nil {
+		return nil, err
+	}
+
+	d.aggregator = newAggregator(d.suite, pub, d.verifiers, commitments, d.t, d.sessionID)
+	// C = F + G
+	d.deals = make([]*Deal, len(d.verifiers))
+	for i := range d.verifiers {
+		fi := f.Eval(i)
+		gi := g.Eval(i)
+		d.deals[i] = &Deal{
+			SessionID:   d.sessionID,
+			SecShare:    fi,
+			RndShare:    gi,
+			Commitments: commitments,
+			T:           uint32(d.t),
+		}
+		if d.deals[i].Signature, err = sign.Schnorr(d.suite, d.long, msgDeal(d.deals[i])); err != nil {
+			return nil, err
+		}
+	}
+	return d, nil
+}
+
+// Deals returns a list of deal for each verifiers.
+func (d *Dealer) Deals() []*Deal {
+	var out = make([]*Deal, len(d.deals))
+	copy(out, d.deals)
+	return out
+}
+
+// ProcessComplaint analyzes the given complaint. If it's a valid complaint,
+// then it returns a Justification. This response must be broadcasted to every
+// participants. If it's an invalid complaint, it returns an error about the
+// complaints. The verifiers will also ignore an invalid Complaint.
+func (d *Dealer) ProcessComplaint(c *Complaint) (*Justification, error) {
+	if err := d.verifyComplaint(c); err != nil {
+		return nil, err
+	}
+	return &Justification{
+		// index is guaranteed to be good because of d.verifyComplaint before
+		Deal:      d.deals[int(c.Index)],
+		Complaint: c,
+	}, nil
+}
+
+// ProcessApprovals looks if the approval is legitimate and stores it. If the
+// approval is not legitimate or has already been received, it returns an error.
+// To knoew whether participants approved the sharing, call `d.EnoughApproval`
```

Comment:
I would rename the function to `ProcessApproval` (singular) because it checks only a single `Approval`, right?

> ProcessApproval  checks the given approval and, if valid, stores it. If the approval is invalid or has been already received, it returns an error. ...